### PR TITLE
Don't leak semaphores on failure

### PR DESCRIPTION
**Pull request by dbt**
Also actually do work in parallel, hopefully.

### CODE REVIEW

File: src/main/kotlin/com/openlattice/users/Auth0SyncTask.kt [hunk 1]
```diff
@@ -69,25 +64,25 @@ class Auth0SyncTask : HazelcastFixedRateTask<Auth0SyncTaskDependencies>, Hazelca
...
-        } catch (ex: Exception) {
-            logger.error("Unable to synchronize users", ex)
-            return
-        }
-
+                }
+                // we want to materialize the list of futures so the work happens in the background.
+                .toList()
+                .forEach {
+                    it.get()
```
Comment (dbt) [inline src/main/kotlin/com/openlattice/users/Auth0SyncTask.kt]: this call is now safe(ish?), it should only throw on non-syncUser problems like interruptedexception.

File: src/main/kotlin/com/openlattice/users/Auth0SyncTask.kt [hunk 2]
```diff
@@ -69,25 +64,25 @@ class Auth0SyncTask : HazelcastFixedRateTask<Auth0SyncTaskDependencies>, Hazelca
...
-                        syncSemaphore.release()
+        ds.userListingService
+                .getUsers()
+                .map {
+                    syncSemaphore.acquire()
+                    ds.executor.submit {
+                        try {
+                            ds.users.syncUser(it)
+                        } catch (ex: Exception) {
+                            logger.error("Unable to synchronize user", ex)
```
Comment (katherinebernstein) [inline src/main/kotlin/com/openlattice/users/Auth0SyncTask.kt]: May be worth logging which user couldn’t be synced

**Comment by dbt** [comment]
I ran this locally for a bit, then checked the logs for runtime. Looks like the syncuser task now takes around 12 seconds in the steady state.

| entries  | start | end | duration |
|----------|-------|-----|----------|
| 1446 | 2020-02-18 14:48:28.664000 | 2020-02-18 14:49:17.431000 | 0:00:48.767000 |
| 1446 | 2020-02-18 14:51:19.955000 | 2020-02-18 14:51:32.308000 | 0:00:12.353000 |
| 1446 | 2020-02-18 14:53:19.971000 | 2020-02-18 14:53:32.534000 | 0:00:12.563000 |
| 1446 | 2020-02-18 14:55:20.278000 | 2020-02-18 14:55:38.223000 | 0:00:17.945000 |
| 1446 | 2020-02-18 14:57:19.939000 | 2020-02-18 14:57:32.187000 | 0:00:12.248000 |
| 1446 | 2020-02-18 14:59:20.018000 | 2020-02-18 14:59:32.203000 | 0:00:12.185000 |
| 1446 | 2020-02-18 15:01:20.063000 | 2020-02-18 15:01:32.460000 | 0:00:12.397000 |
| 1446 | 2020-02-18 15:03:20.007000 | 2020-02-18 15:03:32.075000 | 0:00:12.068000 |
| 1446 | 2020-02-18 15:05:19.972000 | 2020-02-18 15:05:32.246000 | 0:00:12.274000 |
| 1446 | 2020-02-18 15:07:19.900000 | 2020-02-18 15:07:32.154000 | 0:00:12.254000 |

**Comment by dbt** [comment]
This is probably a no-go as written, I'm still seeing `rhizome-offshoot` threads hanging in `Auth0SyncService.syncUser`

**Pull request merged**